### ISSUE TITLE
Provide a helper to start a command, and return exit code, stdout and stderr

### ISSUE THREAD

**Issue by blueyed** [awesomeWM/awesome#185]
It would be nice to have something like the following in `awful.util`:

    exit, stdin, stdout = awful.util.pread2("echo 1; echo 2>&2; exit 3")

The following could be adopted: http://stackoverflow.com/a/16515126/15690

**Comment by psychon** [comment]
Calling `fork()` from lua seems like a really bad idea...
Your example seems to assume that this wraps `bash -c foo` around the code, I wonder if that really is necessary or if at least two different functions for this could be provided
What would this `pread2` function return? Some lgi wrappers around GIO objects?
What happens to `stderr`? The combination of "stdin" and "stdout" seems weird...
If you meant `stderr` instead of `stdin`: I would like to avoid having some blocking code in awful. I have seen to many "my awesome regularly freezes" bug reports that turned out to be blocking calls to some shell scripts which did slow things. Some callback-based mechanism seems nicer...

The short version of all this: Could you tell us more about what exactly this function does? Something like API docs that explain things?

**Comment by actionless** [comment]
i am using such solution, but it's very far from ideal

**Comment by actionless** [comment]
i am using such solution, but it's very far from ideal:
https://github.com/actionless/awesome_config/blob/devel/utils/async.lua

**Comment by Elv13** [comment]
I have one that works
https://github.com/Elv13/awesome-configs/blob/master/utils/fd_async.lua#L294

**Comment by Bew78LesellB** [comment]
I use asyncshell, from lain : 
https://github.com/copycat-killer/lain/blob/master/asyncshell.lua

It doesn't return stderr, nor exit code, but is very usable to fetch data from scripts commands !

**Comment by psychon** [comment]
Implemented in #406 (although that doesn't give you the exit status, do we need that one?)

**Comment by blueyed** [comment]
@psychon 
> doesn't give you the exit status, do we need that one?

Yes, that can be useful, e.g. for things like `ponymix is-muted` etc.
Please re-open in case you agree.
